File: backend/app/schemas/fund.py
"""Fund-related Pydantic models."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class FundSearchResult(BaseModel):
    """Fund search result item."""

    code: str = Field(..., description="基金代码")
    name: str = Field(..., description="基金名称")
    type: str = Field(default="", description="基金类型")

    class Config:
        json_schema_extra = {"example": {"code": "110011", "name": "易方达中小盘混合", "type": "混合型-偏股"}}


class FundBasicInfo(BaseModel):
    """Fund basic information."""

    code: str = Field(..., description="基金代码")
    name: str = Field(..., description="基金名称")
    type: str = Field(default="", description="基金类型")
    company: str = Field(default="", description="基金公司")
    inception_date: date | None = Field(default=None, description="成立日期")
    manager: str = Field(default="", description="基金经理")
    manager_tenure: str = Field(default="", description="基金经理任职时长")
    aum: float | None = Field(default=None, description="基金规模(亿元)")
    nav: float | None = Field(default=None, description="最新单位净值")
    nav_date: date | None = Field(default=None, description="净值日期")
    acc_nav: float | None = Field(default=None, description="累计净值")

    # Fee structure（运作费率来自 F10；认/申购参考 pingzhongdata）
    management_fee: float | None = Field(default=None, description="管理费率(%)，年")
    custody_fee: float | None = Field(default=None, description="托管费率(%)，年")
    sales_service_fee: float | None = Field(default=None, description="销售服务费率(%)，年")
    purchase_fee: float | None = Field(default=None, description="申购费率参考(%)")
    max_subscription_fee_pct: float | None = Field(
        default=None,
        description="认购费率上限参考(%)（募集期）",
    )
    redemption_fee: float | None = Field(
        default=None,
        description="赎回费率（若无法分档则为示意，优先看 redemption_fee_detail）",
    )
    annual_operating_fee_pct: float | None = Field(
        default=None,
        description="年度运作费率合计约值(%)=管理+托管+销售服务",
    )
    redemption_fee_detail: str | None = Field(
        default=None,
        description="赎回费分档文字摘要",
    )
    lockup_note: str | None = Field(
        default=None,
        description="封闭期/持有期/定开等相关说明",
    )
    holding_cost_summary: str | None = Field(
        default=None,
        description="持有成本综合说明（面向用户）",
    )


class NAVDataPoint(BaseModel):
    """Single NAV data point."""

    date: date
    nav: float = Field(..., description="单位净值")
    acc_nav: float | None = Field(default=None, description="累计净值")
    change_pct: float | None = Field(default=None, description="日涨跌幅(%)")


class FundNAVHistory(BaseModel):
    """Fund NAV history data."""

    code: str
    name: str
    data: list[NAVDataPoint]


class FundHolding(BaseModel):
    """Fund holding item."""

    name: str = Field(..., description="持仓名称")
    code: str | None = Field(default=None, description="证券代码")
    ratio: float = Field(..., description="持仓占比(%)")
    shares: float | None = Field(default=None, description="持仓数量(万股)")
    market_value: float | None = Field(default=None, description="市值(万元)")


class FundHoldings(BaseModel):
    """Fund holdings data."""

    code: str
    name: str
    report_date: date | None = Field(default=None, description="报告期")
    stock_holdings: list[FundHolding] = Field(default_factory=list, description="股票持仓")
    bond_holdings: list[FundHolding] = Field(default_factory=list, description="债券持仓")
    industry_allocation: list[FundHolding] = Field(default_factory=list, description="行业分布")


class FundMetrics(BaseModel):
    """Fund performance metrics."""

    code: str
    name: str

    # Return metrics
    return_1m: float | None = Field(default=None, description="近1月收益率(%)")
    return_3m: float | None = Field(default=None, description="近3月收益率(%)")
    return_6m: float | None = Field(default=None, description="近6月收益率(%)")
    return_1y: float | None = Field(default=None, description="近1年收益率(%)")
    return_3y: float | None = Field(default=None, description="近3年收益率(%)")
    return_5y: float | None = Field(default=None, description="近5年收益率(%)")
    return_ytd: float | None = Field(default=None, description="今年以来收益率(%)")
    return_inception: float | None = Field(default=None, description="成立以来收益率(%)")

    # Risk metrics
    max_drawdown: float | None = Field(default=None, description="最大回撤(%)")
    volatility: float | None = Field(default=None, description="年化波动率(%)")
    sharpe_ratio: float | None = Field(default=None, description="夏普比率")

    # Annualized returns
    return_1y_annualized: float | None = Field(default=None, description="近1年年化收益(%)")
    return_3y_annualized: float | None = Field(default=None, description="近3年年化收益(%)")

    risk_labels: list[str] = Field(
        default_factory=list,
        description="基于类型与风险指标的标签（仅供参考）",
    )


class DCASimulateRequest(BaseModel):
    """定投模拟请求。"""

    fund_code: str = Field(..., min_length=6, max_length=10, description="基金代码")
    monthly_amount: float = Field(..., gt=0, le=1_000_000_000, description="每期投入金额（元）")
    start_date: date = Field(..., description="开始日期（含）")
    end_date: date = Field(..., description="结束日期（含）")


class DCAPeriodRow(BaseModel):
    invest_date: date
    nav: float
    amount: float
    shares: float


class DCASummary(BaseModel):
    total_invested: float
    total_shares: float
    periods_count: int
    latest_nav: float
    latest_nav_date: str
    market_value: float
    profit_loss: float
    profit_loss_pct: float


class DCASimulateResponse(BaseModel):
    fund_code: str
    fund_name: str
    periods: list[DCAPeriodRow]
    summary: DCASummary
    disclaimer: str = Field(
        default="模拟未计入申购/赎回费与分红再投，结果仅供参考，不构成投资建议。",
    )


class FundRecommendRequest(BaseModel):
    """规则化选基请求（与前端画像字段对齐）。"""

    risk_level: Literal["conservative", "stable", "balanced", "aggressive", "radical"] = Field(
        ...,
        description="风险偏好",
    )
    horizon: Literal["short", "medium", "long"] = Field(
        ...,
        description="投资期限：短/中/长",
    )
    liquidity: Literal["high", "medium", "low"] = Field(
        ...,
        description="流动性偏好：高/中/低",
    )
    themes: list[str] = Field(
        default_factory=list,
        description="主题 id 列表，如 broad_index、tech；空表示不限主题",
    )
    limit: int = Field(10, ge=3, le=30, description="返回数量")


class RecommendedFund(BaseModel):
    code: str
    name: str
    type: str = ""
    reasons: list[str] = Field(default_factory=list, description="匹配说明")


class FundRecommendResponse(BaseModel):
    funds: list[RecommendedFund]
    profile_summary: str
    disclaimer: str


class FundAnalysisReport(BaseModel):
    """AI-generated fund analysis report."""

    code: str
    name: str
    basic_info: FundBasicInfo
    metrics: FundMetrics
    analysis: str = Field(..., description="AI分析报告内容")
    recommendation: str = Field(default="观望", description="投资建议: 买入/观望/回避")
    generated_at: datetime = Field(default_factory=datetime.now)


class FundCompareRequest(BaseModel):
    """Request model for fund comparison."""

    fund_codes: list[str] = Field(..., min_length=2, max_length=5, description="基金代码列表")
    include_ai: bool = Field(
        default=True,
        description="是否调用 LLM 生成文字对比；为 False 时仅返回指标与雷达图，响应更快",
    )


class FundCompareResult(BaseModel):
    """Fund comparison result."""

    funds: list[FundMetrics]
    radar_data: dict = Field(default_factory=dict, description="雷达图数据")
    analysis: str = Field(default="", description="AI对比分析")
    recommendation: str = Field(default="", description="推荐基金")

File: frontend/src/types/fund.ts
/**
 * Fund-related type definitions
 */

export interface FundSearchResult {
  code: string;
  name: string;
  type: string;
}

export interface FundBasicInfo {
  code: string;
  name: string;
  type: string;
  company: string;
  inception_date: string | null;
  manager: string;
  manager_tenure: string;
  aum: number | null;
  nav: number | null;
  nav_date: string | null;
  acc_nav: number | null;
  /** 年费率，从基金资产计提 */
  management_fee: number | null;
  custody_fee: number | null;
  sales_service_fee: number | null;
  /** 管理+托管+销售服务，年费率合计约值 */
  annual_operating_fee_pct: number | null;
  /** 申购费率参考 */
  purchase_fee: number | null;
  /** 认购费率上限参考（募集期） */
  max_subscription_fee_pct: number | null;
  /** 赎回分档中的最低费率（示意） */
  redemption_fee: number | null;
  /** 赎回分档文字 */
  redemption_fee_detail: string | null;
  /** 封闭/定开/持有期说明 */
  lockup_note: string | null;
  /** 持有成本综合说明 */
  holding_cost_summary: string | null;
}

export interface NAVDataPoint {
  date: string;
  nav: number;
  acc_nav: number | null;
  change_pct: number | null;
}

export interface FundNAVHistory {
  code: string;
  name: string;
  data: NAVDataPoint[];
}

export interface FundHolding {
  name: string;
  code: string | null;
  ratio: number;
  shares: number | null;
  market_value: number | null;
}

export interface FundHoldings {
  code: string;
  name: string;
  report_date: string | null;
  stock_holdings: FundHolding[];
  bond_holdings: FundHolding[];
  industry_allocation: FundHolding[];
}

export interface FundMetrics {
  code: string;
  name: string;
  return_1m: number | null;
  return_3m: number | null;
  return_6m: number | null;
  return_1y: number | null;
  return_3y: number | null;
  return_5y: number | null;
  return_ytd: number | null;
  return_inception: number | null;
  max_drawdown: number | null;
  volatility: number | null;
  sharpe_ratio: number | null;
  return_1y_annualized: number | null;
  return_3y_annualized: number | null;
  /** 后端生成的风险/风格标签 */
  risk_labels?: string[];
}

export interface DCASimulateRequest {
  fund_code: string;
  monthly_amount: number;
  start_date: string;
  end_date: string;
}

export interface DCAPeriodRow {
  invest_date: string;
  nav: number;
  amount: number;
  shares: number;
}

export interface DCASummary {
  total_invested: number;
  total_shares: number;
  periods_count: number;
  latest_nav: number;
  latest_nav_date: string;
  market_value: number;
  profit_loss: number;
  profit_loss_pct: number;
}

export interface DCASimulateResponse {
  fund_code: string;
  fund_name: string;
  periods: DCAPeriodRow[];
  summary: DCASummary;
  disclaimer: string;
}

/** 与后端 FundRecommendRequest 一致 */
export type RecommendRiskLevel = 'conservative' | 'stable' | 'balanced' | 'aggressive' | 'radical';
export type RecommendHorizon = 'short' | 'medium' | 'long';
export type RecommendLiquidity = 'high' | 'medium' | 'low';

export interface FundRecommendRequest {
  risk_level: RecommendRiskLevel;
  horizon: RecommendHorizon;
  liquidity: RecommendLiquidity;
  themes: string[];
  limit: number;
}

export interface RecommendedFund {
  code: string;
  name: string;
  type: string;
  reasons: string[];
}

export interface FundRecommendResponse {
  funds: RecommendedFund[];
  profile_summary: string;
  disclaimer: string;
}

export interface FundAnalysisReport {
  code: string;
  name: string;
  basic_info: FundBasicInfo;
  metrics: FundMetrics;
  analysis: string;
  recommendation: '买入' | '观望' | '回避';
  generated_at: string;
}

export interface RealtimeEstimate {
  code: string;
  name: string;
  nav: number;
  estimate_nav: number;
  estimate_change: number;
  update_time: string;
  nav_date: string;
}

export interface FundCompareRequest {
  fund_codes: string[];
  /** 默认 true；设为 false 仅拉取指标与雷达图，不调用 LLM，速度更快 */
  include_ai?: boolean;
}

export interface RadarSeriesItem {
  name: string;
  code: string;
  values: number[];
}

export interface FundCompareResult {
  funds: FundMetrics[];
  radar_data: {
    dimensions: string[];
    series: RadarSeriesItem[];
  };
  analysis: string;
  recommendation: string;
}
